### PR TITLE
FIX: reconcile recreating records that already exist

### PR DESCRIPTION
**Pull request by nlalonde**
Looking at metrics it appears that new records are being published
every time reconcile is done (currently every 5 minutes).

DDNSSD::Backend publish_record will record a op: "publish",
which ends up in the ResourceRecordStore which chooses not to do an
INSERT because it finds that an identical record already exists.
But it gets recorded in prometheus anyway.

So if we can see that the PTR and TXT records already exist, don't
publish them again.

### CODE REVIEW

File: lib/ddnssd/system.rb [hunk 1]
```diff
@@ -154,21 +154,38 @@ def reconcile_containers(backend)
...
-      @logger.info(progname) { "Found #{our_live_records.length} relevant DNS records." }
+      our_live_records = existing_dns_records.select { |rr| our_record?(rr) }
+      existing_txt_and_ptr_records = existing_dns_records.select do |rr|
+        rr.data.is_a?(Resolv::DNS::Resource::IN::PTR) ||
+          rr.data.is_a?(Resolv::DNS::Resource::IN::TXT)
+      end
+
+      our_desired_records = containers.map { |c| c.dns_records }.flatten(1).uniq
+
+      @logger.info(progname) { "Found #{our_live_records.length} relevant DNS records (excluding TXT and PTR)." }
```
Comment (SamSaffron) [inline lib/ddnssd/system.rb:167]: can we fold the 2 @logger.info's into 1 ?

File: lib/ddnssd/system.rb [hunk 2]
```diff
@@ -154,21 +154,38 @@ def reconcile_containers(backend)
...
       @logger.debug(progname) { (["Relevant DNS records:"] + our_live_records.map { |rr| "#{rr.name} #{rr.ttl} #{rr.type} #{rr.value}" }).join("\n  ") } unless our_live_records.empty?
-      @logger.info(progname) { "Should have #{our_desired_records.length} DNS records." }
+
+      @logger.info(progname) { "Should have #{our_desired_records.length} DNS records (including TXT and PTR)." }
```
Comment (SamSaffron) [inline lib/ddnssd/system.rb:170]: ^^^

File: lib/ddnssd/system.rb [hunk 3]
```diff
@@ -154,21 +154,38 @@ def reconcile_containers(backend)
...
       @logger.debug(progname) { (["Desired DNS records:"] + our_desired_records.map { |rr| "#{rr.name} #{rr.ttl} #{rr.type} #{rr.value}" }).join("\n  ") } unless our_desired_records.empty?
 
       # Delete any of "our" records that are no longer needed
-      @logger.info(progname) { "Deleting #{(our_live_records - our_desired_records).length} DNS records." }
-      (our_live_records - our_desired_records).each { |rr| backend.suppress_record(rr) }
+      records_to_delete = our_live_records - our_desired_records
+      @logger.info(progname) { "Deleting #{records_to_delete.length} DNS records." }
```
Comment (SamSaffron) [inline lib/ddnssd/system.rb:175]: might even fold this one as well, into a single line

**Review by SamSaffron — APPROVED** [review]

**Comment by SamSaffron** [comment]
except for minor logging changes, looks good to me.

**Comment by SamSaffron** [comment]
```
spec/system_spec.rb:476:76: C: RSpec/ReceiveNever: Use not_to receive instead of never.
            expect(mock_backend).to receive(:suppress_record).with(eq(rr)).never
```
robocop is being upset

File: lib/ddnssd/system.rb (hunk 1 above)
Comment (riking) [reply lib/ddnssd/system.rb:167]: also include `existing_dns_records.length` so all 3 numbers are in the log

**Pull request merged**